### PR TITLE
Issue #6420: Strengthen user_pass_rehash()

### LINKED ISSUE #6420: [Security] Fixed: Strengthen one-time login links.
This will address two Drupal 7 improvements:

* Changing email address should invalidate one-time login links
* Increase entropy in the one-time login links (especially in the cases where there's an empty password in the database, if a site uses SSO).

In https://www.drupal.org/project/drupal/issues/3306390:

> Problem/Motivation
> Assume I change email provider or leave a job. I or an attacker requests a password reset link.
> 
> I now change the email address on my drupal account, which I've been logged into on my home computer
> 
> My former employer or someone with access to my old email account can use the pre-existing one-time login link to gain access to my account (or the whole site if I'm an admin).
> 
> Proposed resolution
> Invalidate one-time login links if the email address is changed. The easiest fix is to include the current email address as part of the data in user_pass_rehash

From Drupal 7.100: https://www.drupal.org/node/3409960, https://www.drupal.org/node/3409043 (backport from Drupal 10):

>The problem is that there is significantly less entropy in the inputs to user_pass_rehash() if the user has an empty password field in the database.
>
>Users may have empty passwords if - for example - a site uses a Single Sign On integration to offload authentication.
>
>Improvements could be made to the implementation to reduce the likelihood that an attacker could brute force e.g. a password reset URL if the user has no password stored in Drupal.
>
>It may also be advisable for sites / projects that do not use Drupal's passwords for authentication to avoid leaving the password field empty in the database, but that can be looked at in a (public) follow-up. It wouldn't help much to set the password field to the same value for all users.

### PR DESCRIPTION
**Pull request by herbdool**
Fixes https://github.com/backdrop/backdrop-issues/issues/6420

### CODE REVIEW

**Comment by herbdool** [comment]
I think I found a case where fixing the coding standard breaks things in PHP 5.6:

> PHP Parse error:  syntax error, unexpected ')' in /home/runner/work/backdrop/backdrop/core/modules/user/user.module on line 622

All other PHP versions are fine. And my IDE says it's fine. And me manually counting the open and close brackets says its fine. Sigh. So I'm going to revert that line and let it complain.

File: .cspell/backdrop.dic [hunk 1]
```diff
@@ -1,4 +1,5 @@
 abcdefghijklmnopqrstuvwx
+abcdefghijkmnopqrstuvwxyz
```
Comment (klonos) [inline .cspell/backdrop.dic]: Can we please remove this addition here?

This is non-sensical and only used once in our codebase, so better wrap the line where this is used in `// cspell:disable`/`// cspell:enable` instead.

File: .cspell/backdrop.dic [hunk 2]
```diff
@@ -324,6 +325,7 @@ pagenum
 pagerpieces
 phptemplate
 phpwiki
+plid
```
Comment (klonos) [inline .cspell/backdrop.dic]: This has been added in https://github.com/backdrop/backdrop/pull/4707 so can be removed from this PR (rebasing should fix that).

File: .cspell/backdrop.dic (hunk 1 above)
Comment (herbdool) [reply .cspell/backdrop.dic]: If this one is nonsensical, so is the previous one, which ends at x for some reason.

**Comment by quicksketch** [comment]
Merged into 1.x and 1.27.x. Thanks @herbdool!

**Pull request merged**